cirq/contrib/two_qubit_gates/gate_compilation.py

```diff
@@ -0,0 +1,374 @@
+
+
+class _TabulationStepResult(NamedTuple):
+    # Generated KAK vectors that are uniquely close to at least one mesh point.
+    kept_kaks: List[np.ndarray]
+    # The corresponding single qubit unitaries required to obtain the desired
+    # KAK vectors.
+    kept_cycles: List[Tuple[_SingleQubitGatePair, ...]]
+
+
+def _tabulate_kak_vectors(
+        tabulation: Dict[int, Tuple[_SingleQubitGatePair, ...]],
+        base_gate: np.ndarray,
+        max_dist: float,
+        kak_mesh: np.ndarray,
+        *local_unitary_pairs: _SingleQubitGatePair,
+) -> _TabulationStepResult:
+    """Tabulate KAK vectors from products of local unitaries with a base gate.
+
+    Args:
+        tabulation: Dictionary mapping indices of the KAK vector mesh to the
+            gate products required to achieve the corresponding KAK vector,
+            within some Pauli error.
+        base_gate: The base 2 qubit gate used in the gate product.
+        max_dist: The largest allowed Pauli error between a generated 2Q
+           unitary and a KAK vector mesh point that it is tabulated to.
+        kak_mesh: Sequence of KAK vectors filling the Weyl chamber whose
+            nearest neighbor distance is about 2*max_error.
+        *local_unitary_pairs: 2-Tuples of single qubit unitary tensors, each
+            of shape (N,2,2).
+
+    Returns:
+        The newly tabulated KAK vectors and the local unitaries used to generate
+        them.
+    """
+    shapes = {pair[0].shape for pair in local_unitary_pairs}
+    shapes.update({pair[0].shape for pair in local_unitary_pairs})
+    assert len(shapes) == 1
+    assert len(shapes.pop()) == 3
+
+    # Generate products
+    local_cycles = np.array(
+        [vector_kron(*pairs) for pairs in local_unitary_pairs])
+
+    prods = np.einsum('ab,...bc,cd', base_gate, local_cycles[0], base_gate)
+    for local_cycle in local_cycles[1:]:
+        np.einsum('ab,...bc,...cd', base_gate, local_cycle, prods, out=prods)
+
+    kak_vectors = kak_vector(prods, check_preconditions=False)
+
+    kept_kaks = []
+    kept_cycles = []
+
+    for ind, vec in enumerate(kak_vectors):
+        # dists = KAK_vector_infidelity(vec, KAK_mesh)
+        # The L2 distance is an upper bound to the locally invariant distance,
+        # but it's much faster to compute.
+        dists = np.sqrt(np.sum((kak_mesh - vec)**2, axis=-1))
+        close = (dists < max_dist).nonzero()[0]
+        assert close.shape[0] in (0, 1), f'shape: {close.shape}'
+        cycles_for_gate = tuple(
+            (k_0[ind], k_1[ind]) for k_0, k_1 in local_unitary_pairs)
+
+        kept = False
+        for mesh_ind in close:
+            if mesh_ind not in tabulation:
+                tabulation[mesh_ind] = cycles_for_gate
+                kept = True
+
+        if kept:
+            kept_kaks.append(vec)
+            kept_cycles.append(cycles_for_gate)
+
+    return _TabulationStepResult(kept_kaks, kept_cycles)
+
+
+def gate_product_tabulation(base_gate: np.ndarray,
```

Comment:
nit: I'd consider making the optional arguments named parameters.

Reply:
Done.